File: Leetcode/300/300.go
// O(n^2)
func lengthOfLIS(nums []int) int {
	n := len(nums)
	dp := make([]int, n)
    for i := 0; i < n; i++ {
        dp[i] = 1
    }
	for i := 1; i < n; i++ {
		for j := 0; j < i; j++ {
			if nums[j] < nums[i] {
				if dp[i] < dp[j] + 1 {
					dp[i] = dp[j] + 1
				} 
			}
		}
	}
	ans := 0
	for i := 0; i < n; i++ {
		if dp[i] > ans {
			ans = dp[i]
		}
	}
	return ans
}

File: Leetcode/300/300.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // LIS越长 LIS各元素增长率越小
        int n = (int)nums.size();
        vector<int> lis;
        for (int i = 0; i < n; ++i) {
            auto iter = lower_bound(lis.begin(),lis.end(),nums[i]);
            // 在当前LIS中没有比它大的或者相等的了
            if (iter == lis.end()) {
                lis.emplace_back(nums[i]);
            }
            // 在当前LIS中有比它大的或相等的 替换掉那个大的或相等的
            else {
                *iter = nums[i];
            }
        }
        return (int)lis.size();
    }
};